system/security/securesocket/securesocket.cpp

```diff
@@ -1641,3 +1642,58 @@ SECURESOCKET_API int signCertificate(const char* csr, const char* ca_certificate
 }
 
 }
+
+class CSecureSmartSocketFactory : public CSmartSocketFactory
+{
+public:
+    Owned<ISecureSocketContext> secureContext;
+
+    CSecureSmartSocketFactory(const char *_socklist, bool _retry, unsigned _retryInterval, unsigned _dnsInterval) : CSmartSocketFactory(_socklist, _retry, _retryInterval, _dnsInterval)
+    {
+        secureContext.setown(createSecureSocketContext(ClientSocket));
+    }
+
+    virtual ISmartSocket *connect_timeout(unsigned timeoutms) override
+    {
+        ISocket *sock = nullptr;
+        SmartSocketEndpoint *ss = nullptr;
+        SocketEndpoint ep;
+        ISecureSocket *ssock = nullptr;
+        try
+        {
+            sock = connect_sock(timeoutms, ss, ep);
+            if (!sock)
+                throw createSmartSocketException(0, "securesmartsocket failed to connect");
+
+            ssock = secureContext->createSecureSocket(sock);
+
+            int status = ssock->secure_connect();
```

Comment:
Right, true, needs to set status if secure_socket fails.
So will need a try/catch around:
```
        Owned<ISecureSocket> ssock = secureContext->createSecureSocket(sock.getClear());
        int status = ssock->secure_connect();
        if (status < 0)
        {
            // secure_connect may also DBGLOG() errors ...
            throw MakeStringException(-1, "SSL smartsocket connect fail");
        }
```

to set status.
You could use setStatus, but looks potentially a bit expensive as it scans.
Ok, so maybe do need SmartSocketEndpoint passed back by reference by connect_sock.
Still doesn't need to be exposed in the interface, connect_sock can be protected member and SmartSocketEndpoint still implementation only.